### PR TITLE
Backport qt support minimum to 5.5.1 and associated backports

### PR DESCRIPTION
**Pull request by 10xcryptodev**
Backport bitcoin#18769

Remove the code used to fix a Qt bug in versions before Qt 5.5.0 as described in this link https://bugreports.qt.io/browse/QTBUG-43473

### CODE REVIEW

**Comment by 10xcryptodev** [comment]
@UdjinM6 i can add the backports here up to bitcoin#15393 that uses Qt 5.5.1 as minimum version, do you agree or it should be done later?

**Comment by PastaPastaPasta** [comment]
Is there really any reason for backporting this rn? Normally I try to only backport stuff out of order if I find it is pertinent and timely to backport asap

**Comment by UdjinM6** [comment]
On one hand, it would be nice to narrow it to qt5 and drop all the dead code. On the other, I agree with @PastaPastaPasta that if it's going to cause merge conflicts in the future it's probably not worth it. But it _should_ touch qt only (besides some build scripts maybe) and _should_ mostly remove things so probably _should not_ cause any serious conflicts I guess. Let's maybe try and see how invasive these changes would be if we'd follow that route and then decide?

**Comment by 10xcryptodev** [comment]
@UdjinM6 @PastaPastaPasta 
Added backports:
13458 - Drop Qt4 support
14725 - This one also fix some HiDPI problems
15393 - Requires Qt 5.5.1 as minimum

**Comment by PastaPastaPasta** [comment]
As this is a "non-critical" backport, I'm advocating this get merged in 17

**Comment by UdjinM6** [comment]
Few issues:

1. Mac build fails.
```
qt/macdockiconhandler.mm:36:11: error: no member named 'm_dockMenu' in 'MacDockIconHandler'
    this->m_dockMenu->setAsDockMenu();
```
This was removed in #3463.

2. "14725  - This one also fix some HiDPI problems"
The hidpi related code is really from https://github.com/bitcoin/bitcoin/pull/5228/files#diff-e74648dfe36ec1841b222ca921ad61c0R53. I _think_ we skipped them intentionally because we draw it in a slightly different way. With these changes splashscreen on my mac look like a tiny window.

In both cases, if we don't have some special code for qt5+ already it was either removed for a reason or some intermediary backport is missing, you should _not_ add it from btc manually.

**Comment by UdjinM6** [comment]
Again, 14123 and 16720 were already backported via #3463 recently, you need to rebase on latest develop (and drop these duplicate backports from this branch).

**Comment by 10xcryptodev** [comment]
@UdjinM6 ok i will organize and force push, didn't saw the conflicts with 14123 and 16720 on the other PR

About 14725 i disagree, took a screenshot:
Screenshot from 2020-05-14 20-12-30

without 14725
Screenshot from 2020-05-14 20-09-44

Click on the images to see, this page here don't show well the differences

**Comment by UdjinM6** [comment]
That's not a splash screen though ;) And these changes https://github.com/dashpay/dash/pull/3477/commits/4c82a08931044bee129e505304d6c65b5469bc56#diff-090c60fbf60da129ccb2db498a393f04R572 actually revert this https://github.com/dashpay/dash/pull/3217/files#diff-090c60fbf60da129ccb2db498a393f04R614 part from #3217 which backported the fix from bitcoin and afaik it works for some people. If it doesn't work for you I guess you should also have the same kind of issues with bitcoin-qt on your machine cause current state of things matches https://github.com/bitcoin/bitcoin/blob/master/src/qt/bitcoin.cpp#L428-L434. It also matches the way qt docs describe the way it should be used: `This attribute must be set before QGuiApplication is constructed` https://doc.qt.io/qt-5/qt.html#ApplicationAttribute-enum. There is in fact an open issue in bitcoin core https://github.com/bitcoin/bitcoin/issues/17153 and it looks like they have no solution atm. Maybe in your case it would work if you run it with `QT_AUTO_SCREEN_SCALE_FACTOR=0` env variable (or change `AA_EnableHighDpiScaling` to `AA_DisableHighDpiScaling` in code and btw, you can read a bit more about hidpi in qt here https://doc.qt.io/qt-5/highdpi.html) but in any case it's a separate issue and these changes do not belong to "bump min qt to 5.2" PR.

**Comment by 10xcryptodev** [comment]
@UdjinM6 thanks for the info, also saw that the display that i tested is high resolution but not HiDPI
```
screen #0:
  dimensions:    3840x2160 pixels (1016x572 millimeters)
  resolution:    96x96 dots per inch
```
So it seems that there are some problems with scale/resolution and DPI, i will do some investigations in the future, with a HiDPI display too

I will remove the 14725 and organize the branch

**Comment by UdjinM6** [comment]
Original 14725 itself is fine though, it had no changes like ones I pointed to in your backport. IMO it should not be dropped, it just should be backported correctly, without touching/adding unrelated code.

File: configure.ac [hunk 1]
```diff
@@ -856,7 +856,7 @@ fi
 BITCOIN_QT_INIT
 
 dnl sets $bitcoin_enable_qt, $bitcoin_enable_qt_test, $bitcoin_enable_qt_dbus
-BITCOIN_QT_CONFIGURE([$use_pkgconfig], [qt5])
+BITCOIN_QT_CONFIGURE([$use_pkgconfig])
```
Comment (UdjinM6) [inline configure.ac:859]: missing changes for `bitcoin_qt_got_major_vers` (line 1450)

File: src/qt/dash.cpp [hunk 2]
```diff
@@ -589,13 +566,7 @@ int main(int argc, char *argv[])
...
     // Do not refer to data directory yet, this can be overridden by Intro::pickDataDirectory
 
-    /// 2. Basic Qt initialization (not dependent on parameters or configuration)
-#if QT_VERSION < 0x050000
-    // Internal string conversion is all UTF-8
-    QTextCodec::setCodecForTr(QTextCodec::codecForName("UTF-8"));
-    QTextCodec::setCodecForCStrings(QTextCodec::codecForTr());
-#endif
-
+    /// 1. Basic Qt initialization (not dependent on parameters or configuration)
```
Comment (UdjinM6) [inline src/qt/dash.cpp:566]:
```suggestion
    /// 2. Basic Qt initialization (not dependent on parameters or configuration)
```

File: build-aux/m4/bitcoin_qt.m4 [hunk 3]
```diff
@@ -291,7 +273,7 @@ AC_DEFUN([_BITCOIN_QT_CHECK_QT5],[
       #endif
     ]],
     [[
-      #if QT_VERSION < 0x050000
+      #if QT_VERSION < 0x050501
```
Comment (UdjinM6) [inline build-aux/m4/bitcoin_qt.m4:276]: missing changes for `QTXCBQPA`

**Comment by 10xcryptodev** [comment]
@UdjinM6 done :+1:

File: configure.ac (hunk 1 above)
Comment (UdjinM6) [reply configure.ac:859]: still not fixed https://github.com/bitcoin/bitcoin/pull/13458/files#diff-67e997bcfdac55191033d57a16d1408aL1425

File: build-aux/m4/bitcoin_qt.m4 (hunk 3 above)
Comment (UdjinM6) [reply build-aux/m4/bitcoin_qt.m4:276]: still missing https://github.com/bitcoin/bitcoin/pull/15393/files#diff-ab26cccb62e7a807f1dce314412a5e17R359

**Comment by 10xcryptodev** [comment]
@UdjinM6 done again :) :+1:

**Comment by 10xcryptodev** [comment]
@PastaPastaPasta updated with bitcoin#18769 in the same way, thanks

**Comment by PastaPastaPasta** [comment]
Did you add another backport? I thought this PR had three backports + your commit, which then would become 4 backports, but I see five here. Am I misremembering?

**Pull request merged**